core/src/main/scala/neotypes/DeferredQuery.scala

```diff
@@ -43,51 +41,52 @@ case class DeferredQuery[T](query: String, params: Map[String, Any] = Map()) {
   def stream[S[_], F[_]](tx: Transaction[F])(implicit rm: ResultMapper[T], sb: Stream[S, F]): S[T] =
     tx.stream(query, params)
 
-  def withParams(params: Map[String, Any]): DeferredQuery[T] = copy(params = this.params ++ params)
-
+  def withParams(params: Map[String, Any]): DeferredQuery[T] =
+    copy(params = this.params ++ params)
 }
 
-class DeferredQueryBuilder(private[neotypes] val parts: Seq[Part]) {
-  protected[neotypes] lazy val rawQuery: String = parts.zipWithIndex.map {
-    case (Query(part), _) => part
-    case (Param(_), index) => "$" + PARAMETER_NAME_PREFIX + (index / 2 + 1)
-  }.mkString("")
+private[neotypes] class DeferredQueryBuilder(private[neotypes] val parts: Seq[DeferredQueryBuilder.Part]) {
+  def this(query: String) =
+    this(Seq(DeferredQueryBuilder.Query(query)))
+
+  import DeferredQueryBuilder.{PARAMETER_NAME_PREFIX, Param, Part, Query}
 
-  protected[neotypes] lazy val params: Map[String, Any] = parts.collect {
-    case Param(value) => value
-  }.zipWithIndex.map {
-    case (value, index) => PARAMETER_NAME_PREFIX + (index + 1) -> value
-  }.toMap
+  protected[neotypes] lazy val rawQuery: String =
+    parts.iterator.zipWithIndex.map {
+      case (Query(part), _)  => part
+      case (Param(_), index) => s"$$${PARAMETER_NAME_PREFIX}${index / 2 + 1}"
+    }.mkString
 
-  def this(query: String) = this(Seq(Query(query)))
+  protected[neotypes] lazy val params: Map[String, Any] =
+    parts.iterator.collect {
+      case Param(value) => value
+    }.zipWithIndex.map {
+      case (value, index) => s"${PARAMETER_NAME_PREFIX}${index + 1}" -> value
+    }.toMap
 
-  def query[T]: DeferredQuery[T] = {
+  def query[T]: DeferredQuery[T] =
     DeferredQuery(rawQuery, params)
-  }
 
-  def +(other: DeferredQueryBuilder): DeferredQueryBuilder = {
+  def +(other: DeferredQueryBuilder): DeferredQueryBuilder =
     new DeferredQueryBuilder(
-      (parts.slice(0, parts.size - 1) :+ parts.last.asInstanceOf[Query].merge(other.parts.head.asInstanceOf[Query])) ++ other.parts.tail
+      (parts.init :+ parts.last.asInstanceOf[Query].merge(other.parts.head.asInstanceOf[Query])) ++ other.parts.tail
     )
-  }
 
-  def +(other: String): DeferredQueryBuilder = {
+  def +(other: String): DeferredQueryBuilder =
```

Comment:
Do you believe is useful to have these concat functions?

IMHO, I would be better to remove this **Builder** and just have the `c` _interpolator_ return a **PreDeferredQuery** _(I am bad for names)_ which simply has the `query` method.
And, also create an implicit class for strings with the `query` method too.

My reasoning is that, I do not believe people will be concatenating larger queries, but instead will use a multi-line string.
And, if for some reason, they are building the query dynamically, the will be using plain strings anyways, so they may just concat the strings first and the call `query` on the result.

Reply:
I just finished with the new builder :D